pkg/pushsync/pushsync.go

```diff
@@ -219,80 +224,187 @@ func (ps *PushSync) PushChunkToClosest(ctx context.Context, ch swarm.Chunk) (*Re
 	span, _, ctx := ps.tracer.StartSpanFromContext(ctx, "pushsync-push", ps.logger, opentracing.Tag{Key: "address", Value: ch.Address().String()})
 	defer span.Finish()
 
-	peer, err := ps.peerSuggester.ClosestPeer(ch.Address())
-	if err != nil {
-		if errors.Is(err, topology.ErrWantSelf) {
-			// this is to make sure that the sent number does not diverge from the synced counter
-			t, err := ps.tagg.Get(ch.TagID())
-			if err == nil && t != nil {
-				err = t.Inc(tags.StateSent)
-				if err != nil {
-					return nil, err
+	var (
+		skipPeers []swarm.Address
+		lastErr   error
+	)
+
+	for i := 0; i < maxPeers; i++ {
+		select {
+		case <-ctx.Done():
+			return nil, ctx.Err()
+		default:
+		}
+
+		// find next closes peer
+		var (
+			peer swarm.Address
+			err  error
+		)
+
+		if i == 0 {
+			peer, err = ps.peerSuggester.ClosestPeer(ch.Address())
+			if err != nil {
+				if errors.Is(err, topology.ErrNotFound) {
+					// NOTE: needed for tests
+					continue
 				}
+
+				if errors.Is(err, topology.ErrWantSelf) {
+					// this is to make sure that the sent number does not diverge from the synced counter
+					t, err := ps.tagg.Get(ch.TagID())
+					if err == nil && t != nil {
+						err = t.Inc(tags.StateSent)
+						if err != nil {
+							return nil, err
+						}
+					}
+
+					// if you are the closest node return a receipt immediately
+					return &Receipt{
+						Address: ch.Address(),
+					}, nil
+				}
+
+				return nil, fmt.Errorf("closest peer: %w", err)
+			}
+		} else {
+			peer, err = ps.closestPeer(ch.Address(), skipPeers)
+			if err != nil {
+				return nil, fmt.Errorf("closest peer: %w", err)
 			}
+		}
 
-			// if you are the closest node return a receipt immediately
-			return &Receipt{
-				Address: ch.Address(),
-			}, nil
+		// save found peer (to be skipped if there is some error with him)
+		skipPeers = append(skipPeers, peer)
+
+		// compute the price we pay for this receipt and reserve it for the rest of this function
+		receiptPrice := ps.pricer.PeerPrice(peer, ch.Address())
+		err = ps.accounting.Reserve(ctx, peer, receiptPrice)
+		if err != nil {
+			return nil, fmt.Errorf("reserve balance for peer %s: %w", peer.String(), err)
 		}
-		return nil, fmt.Errorf("closest peer: %w", err)
-	}
+		defer ps.accounting.Release(peer, receiptPrice)
 
-	// compute the price we pay for this receipt and reserve it for the rest of this function
-	receiptPrice := ps.pricer.PeerPrice(peer, ch.Address())
-	err = ps.accounting.Reserve(ctx, peer, receiptPrice)
-	if err != nil {
-		return nil, fmt.Errorf("reserve balance for peer %s: %w", peer.String(), err)
-	}
-	defer ps.accounting.Release(peer, receiptPrice)
+		streamer, err := ps.streamer.NewStream(ctx, peer, nil, protocolName, protocolVersion, streamName)
+		if err != nil {
+			lastErr = fmt.Errorf("new stream for peer %s: %w", peer.String(), err)
+			ps.logger.Debugf("pushsync-push: %w", lastErr)
+			continue
+		}
+		defer func() { go streamer.FullClose() }()
```

Comment:
If there are errors after this defer, there may be multiple FullClose calls. I do not think that it is a problem, but it is not necessary.